ksqldb-engine/src/main/java/io/confluent/ksql/physical/scalablepush/ScalablePushRegistry.java

```diff
@@ -364,4 +457,129 @@ private void runLatest(final LatestConsumer latestConsumerToRun) {
   private String getLatestConsumerGroupId() {
     return sourceApplicationId + LATEST_CONSUMER_GROUP_SUFFIX;
   }
+
+  private synchronized void catchupAssignmentUpdater(
+      final Collection<TopicPartition> newAssignment
+  ) {
+    for (CatchupConsumer catchupConsumer : catchupConsumers.values()) {
+      catchupConsumer.newAssignment(newAssignment);
+    }
+  }
+
+  /**
+   * Creates the latest consumer and its underlying kafka consumer.
+   * @param processingQueue The queue on which to send an error if anything goes wrong
+   * @return The new LatestConsumer
+   */
+  private CatchupConsumer createCatchupConsumer(
+      final ProcessingQueue processingQueue,
+      final PushOffsetRange offsetRange,
+      final String consumerGroup
+  ) {
+    KafkaConsumer<Object, GenericRow> consumer = null;
+    CatchupConsumer catchupConsumer = null;
+    try {
+      consumer = kafkaConsumerFactory.create(
+          ksqlTopic, logicalSchema, serviceContext, consumerProperties, ksqlConfig,
+          consumerGroup);
+      catchupConsumer = catchupConsumerFactory.create(ksqlTopic.getKafkaTopicName(), isWindowed(),
+          logicalSchema, consumer, latestConsumer::get, catchupCoordinator, offsetRange,
+          Clock.systemUTC());
+      return catchupConsumer;
+    } catch (Exception e) {
+      LOG.error("Couldn't create catchup consumer", e);
+      processingQueue.onError();
+      // We're not supposed to block here, but if it fails here, hopefully it can immediately close.
+      if (consumer != null) {
+        consumer.close();
+      }
+      throw e;
+    }
+  }
+
+  /**
+   * Starts the catchup consumer
+   * @param processingQueue The queue to register with the catchup consumer.
+   * @param catchupMetadata The catchup metadata
+   */
+  private synchronized void startCatchup(
+      final ProcessingQueue processingQueue,
+      final CatchupMetadata catchupMetadata
+  ) {
+    final CatchupConsumer catchupConsumer = createCatchupConsumer(processingQueue,
+        catchupMetadata.getPushOffsetRange(),
+        catchupMetadata.getCatchupConsumerGroup());
+    catchupConsumer.register(processingQueue);
+    catchupConsumers.put(processingQueue.getQueryId(), catchupConsumer);
+    catchupCounter.incrementAndGet();
+    executorServiceCatchup.submit(() -> runCatchup(catchupConsumer, processingQueue));
+  }
+
+  private void runCatchup(
+      final CatchupConsumer catchupConsumerToRun,
+      final ProcessingQueue processingQueue
+  ) {
+    try (CatchupConsumer catchupConsumer = catchupConsumerToRun) {
+      catchupConsumer.run();
+    } catch (Throwable t) {
+      LOG.error("Got error while running catchup", t);
+      catchupConsumerToRun.onError();
+    } finally {
+      catchupConsumerToRun.unregister(processingQueue);
+      catchupConsumers.remove(processingQueue.getQueryId());
+      catchupCounter.decrementAndGet();
+      // If things ended exceptionally, stop latest
+      stopLatestConsumerOnLastRequest();
```

Comment:
these are littered throughout the code base. is it possible to just have the latest consumer stop itself if there are no requests registered so we don't have to do that bookkeeping?

Reply:
The issue is that it needs to have latest running even if the only queue is for a catchup.  You might want to stop if if there's an error creating the catchup or if you unregister the last queue.  Not sure there's a way around it.